File: Source/Samples/Jacobi.Vst.Samples.Host/HostCommandStub.cs
using Jacobi.Vst.Core;
using Jacobi.Vst.Core.Host;
using System;

namespace Jacobi.Vst.Samples.Host
{
    /// <summary>
    /// The HostCommandStub class represents the part of the host that a plugin can call.
    /// </summary>
    class HostCommandStub : IVstHostCommandStub
    {
        public HostCommandStub()
        {
            Commands = new HostCommands(this);
        }

        /// <summary>
        /// Raised when one of the methods is called.
        /// </summary>
        public event EventHandler<PluginCalledEventArgs> PluginCalled;

        private void RaisePluginCalled(string message)
        {
            PluginCalled?.Invoke(this, new PluginCalledEventArgs(message));
        }

        #region IVstHostCommandsStub Members

        /// <inheritdoc />
        public IVstPluginContext PluginContext { get; set; }

        public IVstHostCommands20 Commands { get; private set; }

        #endregion

        private class HostCommands : IVstHostCommands20
        {
            private readonly HostCommandStub _cmdStub;

            public HostCommands(HostCommandStub cmdStub)
            {
                _cmdStub = cmdStub;
            }

            #region IVstHostCommands20 Members

            /// <inheritdoc />
            public bool BeginEdit(int index)
            {
                _cmdStub.RaisePluginCalled("BeginEdit(" + index + ")");

                return false;
            }

            /// <inheritdoc />
            public Jacobi.Vst.Core.VstCanDoResult CanDo(string cando)
            {
                _cmdStub.RaisePluginCalled("CanDo(" + cando + ")");
                return Jacobi.Vst.Core.VstCanDoResult.Unknown;
            }

            /// <inheritdoc />
            public bool CloseFileSelector(Jacobi.Vst.Core.VstFileSelect fileSelect)
            {
                _cmdStub.RaisePluginCalled("CloseFileSelector(" + fileSelect.Command + ")");
                return false;
            }

            /// <inheritdoc />
            public bool EndEdit(int index)
            {
                _cmdStub.RaisePluginCalled("EndEdit(" + index + ")");
                return false;
            }

            /// <inheritdoc />
            public Jacobi.Vst.Core.VstAutomationStates GetAutomationState()
            {
                _cmdStub.RaisePluginCalled("GetAutomationState()");
                return Jacobi.Vst.Core.VstAutomationStates.Off;
            }

            /// <inheritdoc />
            public int GetBlockSize()
            {
                _cmdStub.RaisePluginCalled("GetBlockSize()");
                return 1024;
            }

            /// <inheritdoc />
            public string GetDirectory()
            {
                _cmdStub.RaisePluginCalled("GetDirectory()");
                return null;
            }

            /// <inheritdoc />
            public int GetInputLatency()
            {
                _cmdStub.RaisePluginCalled("GetInputLatency()");
                return 0;
            }

            /// <inheritdoc />
            public Jacobi.Vst.Core.VstHostLanguage GetLanguage()
            {
                _cmdStub.RaisePluginCalled("GetLanguage()");
                return Jacobi.Vst.Core.VstHostLanguage.NotSupported;
            }

            /// <inheritdoc />
            public int GetOutputLatency()
            {
                _cmdStub.RaisePluginCalled("GetOutputLatency()");
                return 0;
            }

            /// <inheritdoc />
            public Jacobi.Vst.Core.VstProcessLevels GetProcessLevel()
            {
                _cmdStub.RaisePluginCalled("GetProcessLevel()");
                return Jacobi.Vst.Core.VstProcessLevels.Unknown;
            }

            /// <inheritdoc />
            public string GetProductString()
            {
                _cmdStub.RaisePluginCalled("GetProductString()");
                return "VST.NET";
            }

            /// <inheritdoc />
            public float GetSampleRate()
            {
                _cmdStub.RaisePluginCalled("GetSampleRate()");
                return 44.8f;
            }

            /// <inheritdoc />
            public Jacobi.Vst.Core.VstTimeInfo GetTimeInfo(Jacobi.Vst.Core.VstTimeInfoFlags filterFlags)
            {
                _cmdStub.RaisePluginCalled("GetTimeInfo(" + filterFlags + ")");
                return null;
            }

            /// <inheritdoc />
            public string GetVendorString()
            {
                _cmdStub.RaisePluginCalled("GetVendorString()");
                return "Jacobi Software";
            }

            /// <inheritdoc />
            public int GetVendorVersion()
            {
                _cmdStub.RaisePluginCalled("GetVendorVersion()");
                return 1000;
            }

            /// <inheritdoc />
            public bool IoChanged()
            {
                _cmdStub.RaisePluginCalled("IoChanged()");
                return false;
            }

            /// <inheritdoc />
            public bool OpenFileSelector(Jacobi.Vst.Core.VstFileSelect fileSelect)
            {
                _cmdStub.RaisePluginCalled("OpenFileSelector(" + fileSelect.Command + ")");
                return false;
            }

            /// <inheritdoc />
            public bool ProcessEvents(Jacobi.Vst.Core.VstEvent[] events)
            {
                _cmdStub.RaisePluginCalled("ProcessEvents(" + events.Length + ")");
                return false;
            }

            /// <inheritdoc />
            public bool SizeWindow(int width, int height)
            {
                _cmdStub.RaisePluginCalled("SizeWindow(" + width + ", " + height + ")");
                return false;
            }

            /// <inheritdoc />
            public bool UpdateDisplay()
            {
                _cmdStub.RaisePluginCalled("UpdateDisplay()");
                return false;
            }

            #endregion

            #region IVstHostCommands10 Members

            /// <inheritdoc />
            public int GetCurrentPluginID()
            {
                _cmdStub.RaisePluginCalled("GetCurrentPluginID()");
                return _cmdStub.PluginContext.PluginInfo.PluginID;
            }

            /// <inheritdoc />
            public int GetVersion()
            {
                _cmdStub.RaisePluginCalled("GetVersion()");
                return 1000;
            }

            /// <inheritdoc />
            public void ProcessIdle()
            {
                _cmdStub.RaisePluginCalled("ProcessIdle()");
            }

            /// <inheritdoc />
            public void SetParameterAutomated(int index, float value)
            {
                _cmdStub.RaisePluginCalled("SetParameterAutomated(" + index + ", " + value + ")");
            }

            #endregion
        }
    }

    /// <summary>
    /// Event arguments used when one of the mehtods is called.
    /// </summary>
    class PluginCalledEventArgs : EventArgs
    {
        /// <summary>
        /// Constructs a new instance with a <paramref name="message"/>.
        /// </summary>
        /// <param name="message"></param>
        public PluginCalledEventArgs(string message)
        {
            Message = message;
        }

        /// <summary>
        /// Gets the message.
        /// </summary>
        public string Message { get; private set; }
    }
}

File: Source.Core/Code/Jacobi.Vst.Interop/Plugin/HostCommandStub.cpp
#include "pch.h"
#include "HostCommandStub.h"
#include "../TypeConverter.h"
#include "../UnmanagedString.h"
#include "../Properties\Resources.h"
#include "../Utils.h"

namespace Jacobi {
namespace Vst {
namespace Plugin {
namespace Interop {

// Creates a new instance based on a native callback function pointer.
HostCommandStub::HostCommandStub(::Vst2HostCommand hostCommand)
{
	if(hostCommand == NULL)
	{
		throw gcnew System::ArgumentNullException("hostCallback");
	}

	_hostCommand = hostCommand;
	_pluginInfo = NULL;

	_timeInfo = gcnew Jacobi::Vst::Core::VstTimeInfo();
	_traceCtx = gcnew Jacobi::Vst::Core::Diagnostics::TraceContext(
		Jacobi::Vst::Interop::Utils::GetPluginName() + ".Plugin.HostCommandStub", Jacobi::Vst::Core::Plugin::IVstHostCommandStub::typeid);
}

// destructor. See Finalizer
HostCommandStub::~HostCommandStub()
{
	this->!HostCommandStub();
}

// Finalizer deletes the Vst2Plugin instance
HostCommandStub::!HostCommandStub()
{
	_hostCommand = NULL;
	
	if(_pluginInfo != NULL)
	{
		delete _pluginInfo;
		_pluginInfo = NULL;
	}
}

// IVstHostCommandStub
// Updates the passed pluginInfo in with the host.
System::Boolean HostCommandStub::UpdatePluginInfo(Jacobi::Vst::Core::Plugin::VstPluginInfo^ pluginInfo)
{
	if(_pluginInfo && pluginInfo)
	{
		// overwrite the Vst2Plugin values with the new values supplied by pluginInfo.
		_pluginInfo->parameterCount = pluginInfo->ParameterCount;
		_pluginInfo->programCount = pluginInfo->ProgramCount;

		if (_pluginInfo->inputCount != pluginInfo->AudioInputCount ||
			_pluginInfo->outputCount != pluginInfo->AudioOutputCount ||
			_pluginInfo->startupDelay != pluginInfo->InitialDelay)
		{
			_pluginInfo->inputCount = pluginInfo->AudioInputCount;
			_pluginInfo->outputCount = pluginInfo->AudioOutputCount;
			_pluginInfo->startupDelay = pluginInfo->InitialDelay;

			return IoChanged();
		}

		return true;
	}

	return false;
}

// IVstHostCommands10
void HostCommandStub::SetParameterAutomated(System::Int32 index, System::Single value)
{
	ThrowIfNotInitialized();

	CallHost(Vst2HostCommands::Automate, index, 0, 0, value);
}

System::Int32 HostCommandStub::GetVersion()
{
	int32_t version = (::int32_t)CallHost(Vst2HostCommands::Version, 0, 0, 0, 0);

	if(version == 0)	// old host
		version = 1;

	return version;
}

System::Int32 HostCommandStub::GetCurrentPluginID()
{
	return (::int32_t)CallHost(Vst2HostCommands::CurrentId, 0, 0, 0, 0);
}

void HostCommandStub::ProcessIdle()
{
	ThrowIfNotInitialized();

	CallHost(Vst2HostCommands::Idle, 0, 0, 0, 0);
}

// IVstHostCommands20
Jacobi::Vst::Core::VstTimeInfo^ HostCommandStub::GetTimeInfo(Jacobi::Vst::Core::VstTimeInfoFlags filterFlags)
{
	ThrowIfNotInitialized();

	::Vst2TimeInfo* pTimeInfo = (::Vst2TimeInfo*)
		CallHost(Vst2HostCommands::GetTime, 0, safe_cast<int32_t>(filterFlags), 0, 0);

	Jacobi::Vst::Interop::TypeConverter::ToManagedTimeInfo(_timeInfo, pTimeInfo);

	return _timeInfo;
}

System::Boolean HostCommandStub::ProcessEvents(array<Jacobi::Vst::Core::VstEvent^>^ events)
{
	ThrowIfNotInitialized();

	::Vst2Events* pEvents = Jacobi::Vst::Interop::TypeConverter::AllocUnmanagedEvents(events);

	try
	{
		return (CallHost(Vst2HostCommands::ProcessEvents, 0, 0, pEvents, 0) != 0);
	}
	finally
	{
		Jacobi::Vst::Interop::TypeConverter::DeleteUnmanagedEvents(pEvents);
	}

	return false;
}

System::Boolean HostCommandStub::IoChanged()
{
	return (CallHost(Vst2HostCommands::IoChanged, 0, 0, 0, 0) != 0);
}

System::Boolean HostCommandStub::SizeWindow(System::Int32 width, System::Int32 height)
{
	return (CallHost(Vst2HostCommands::SizeWindow, width, height, 0, 0) != 0);
}

System::Single HostCommandStub::GetSampleRate()
{
	return safe_cast<System::Single>(CallHost(Vst2HostCommands::GetSampleRate, 0, 0, 0, 0));
}

System::Int32 HostCommandStub::GetBlockSize()
{
	return (::int32_t)CallHost(Vst2HostCommands::GetBlockSize, 0, 0, 0, 0);
}

System::Int32 HostCommandStub::GetInputLatency()
{
	return (::int32_t)CallHost(Vst2HostCommands::GetInputLatency, 0, 0, 0, 0);
}

System::Int32 HostCommandStub::GetOutputLatency()
{
	return (::int32_t)CallHost(Vst2HostCommands::GetOutputLatency, 0, 0, 0, 0);
}

Jacobi::Vst::Core::VstProcessLevels HostCommandStub::GetProcessLevel()
{
	return safe_cast<Jacobi::Vst::Core::VstProcessLevels>(
		CallHost(Vst2HostCommands::GetCurrentProcessLevel, 0, 0, 0, 0));
}

Jacobi::Vst::Core::VstAutomationStates HostCommandStub::GetAutomationState()
{
	return safe_cast<Jacobi::Vst::Core::VstAutomationStates>(
		CallHost(Vst2HostCommands::GetAutomationState, 0, 0, 0, 0));
}

System::String^ HostCommandStub::GetVendorString()
{
	Jacobi::Vst::Interop::UnmanagedString pText(Vst2MaxVendorStrLen + 1);

	if(CallHost(Vst2HostCommands::VendorGetString, 0, 0, pText, 0) != 0)
	{
		return Jacobi::Vst::Interop::TypeConverter::CharToString(pText);
	}

	return nullptr;
}

System::String^ HostCommandStub::GetProductString()
{
	Jacobi::Vst::Interop::UnmanagedString pText(Vst2MaxProductStrLen + 1);

	if(CallHost(Vst2HostCommands::ProductGetString, 0, 0, pText, 0) != 0)
	{
		return Jacobi::Vst::Interop::TypeConverter::CharToString(pText);
	}

	return nullptr;
}

System::Int32 HostCommandStub::GetVendorVersion()
{
	return (::int32_t)CallHost(Vst2HostCommands::VendorGetVersion, 0, 0, 0, 0);
}

Jacobi::Vst::Core::VstCanDoResult HostCommandStub::CanDo(System::String^ cando)
{
	Jacobi::Vst::Interop::UnmanagedString pText(Jacobi::Vst::Core::Constants::MaxCanDoLength + 1);

	Jacobi::Vst::Interop::TypeConverter::StringToChar(cando, pText, Jacobi::Vst::Core::Constants::MaxCanDoLength);

	return safe_cast<Jacobi::Vst::Core::VstCanDoResult>(
		CallHost(Vst2HostCommands::CanDo, 0, 0, pText, 0));
}

Jacobi::Vst::Core::VstHostLanguage HostCommandStub::GetLanguage()
{
	return safe_cast<Jacobi::Vst::Core::VstHostLanguage>(
		CallHost(Vst2HostCommands::GetLanguage, 0, 0, 0, 0));
}

System::String^ HostCommandStub::GetDirectory()
{
	return Jacobi::Vst::Interop::TypeConverter::CharToString((char*)
		CallHost(Vst2HostCommands::GetDirectory, 0, 0, 0, 0));
}

System::Boolean HostCommandStub::UpdateDisplay()
{
	ThrowIfNotInitialized();

	return (CallHost(Vst2HostCommands::UpdateDisplay, 0, 0, 0, 0) != 0);
}

System::Boolean HostCommandStub::BeginEdit(System::Int32 index)
{
	ThrowIfNotInitialized();
	
	return (CallHost(Vst2HostCommands::EditBegin, index, 0, 0, 0) != 0);
}

System::Boolean HostCommandStub::EndEdit(System::Int32 index)
{
	ThrowIfNotInitialized();
	
	return (CallHost(Vst2HostCommands::EditEnd, index, 0, 0, 0) != 0);
}

System::Boolean HostCommandStub::OpenFileSelector(Jacobi::Vst::Core::VstFileSelect^ fileSelect)
{
	ThrowIfNotInitialized();

	if(fileSelect->Reserved != System::IntPtr::Zero)
	{
		throw gcnew System::InvalidOperationException(
			Jacobi::Vst::Interop::Properties::Resources::HostCommandStub_VstFileSelectAlreadyInitialized);
	}

	::Vst2FileSelect* pFileSelect = Jacobi::Vst::Interop::TypeConverter::AllocUnmanagedFileSelect(fileSelect);

	try
	{
		System::Boolean succeeded = (CallHost(Vst2HostCommands::FileSelectorOpen, 0, 0, pFileSelect, 0) != 0);

		if(succeeded)
		{
			// update the managed type with the users selections.
			Jacobi::Vst::Interop::TypeConverter::UpdateManagedFileSelect(fileSelect, pFileSelect);
		}

		return succeeded;
	}
	catch(...)
	{
		Jacobi::Vst::Interop::TypeConverter::DeleteUnmanagedFileSelect(pFileSelect);

		_traceCtx->WriteEvent(System::Diagnostics::TraceEventType::Error, 
			"Error in Jacobi.Vst.Interop.Plugin.HostCommandStub.OpenFileSelector.");
	}

	return false;
}

System::Boolean HostCommandStub::CloseFileSelector(Jacobi::Vst::Core::VstFileSelect^ fileSelect)
{
	ThrowIfNotInitialized();

	::Vst2FileSelect* pFileSelect = (::Vst2FileSelect*)fileSelect->Reserved.ToPointer();

	if(pFileSelect == NULL)
	{
		throw gcnew System::InvalidOperationException(
			Jacobi::Vst::Interop::Properties::Resources::HostCommandStub_VstFileSelectAlreadyDisposed);
	}

	try
	{
		System::Boolean succeeded = (CallHost(Vst2HostCommands::FileSelectorClose, 0, 0, pFileSelect, 0) != 0);

		return succeeded;
	}
	finally
	{
		Jacobi::Vst::Interop::TypeConverter::DeleteUnmanagedFileSelect(pFileSelect);
		fileSelect->Reserved = System::IntPtr::Zero;
	}
}

//
// Legacy VST 2.4 methods
//

// IVstPluginCommandsLegacy10
System::Boolean HostCommandStub::PinConnected(System::Int32 connectionIndex, System::Boolean output)
{
	// Note: retval 0 = true
	return (CallHost(Vst2HostCommands::PinConnected, connectionIndex, safe_cast<Vst2IntPtr>(output), 0, 0) == 0);
}

// IVstPluginCommandsLegacy20
System::Boolean HostCommandStub::WantMidi()
{
	return (CallHost(Vst2HostCommands::WantMidi, 0, 0, 0, 0) != 0);
}

System::Boolean HostCommandStub::SetTime(Jacobi::Vst::Core::VstTimeInfo^ timeInfo, Jacobi::Vst::Core::VstTimeInfoFlags filterFlags)
{
	::Vst2TimeInfo* pTimeInfo = Jacobi::Vst::Interop::TypeConverter::AllocUnmanagedTimeInfo(timeInfo);

	try
	{
		return (CallHost(Vst2HostCommands::SetTime, 0, safe_cast<Vst2IntPtr>(filterFlags), pTimeInfo, 0) != 0);
	}
	finally
	{
		delete pTimeInfo;
	}
}

System::Int32 HostCommandStub::GetTempoAt(System::Int32 sampleIndex) // bpm * 10000
{
	return safe_cast<System::Int32>(
		CallHost(Vst2HostCommands::TempoAt, 0, safe_cast<Vst2IntPtr>(sampleIndex), 0, 0));
}

System::Int32 HostCommandStub::GetAutomatableParameterCount()
{
	return safe_cast<System::Int32>(
		CallHost(Vst2HostCommands::GetAutomatableParameterCount, 0, 0, 0, 0));
}

System::Int32 HostCommandStub::GetParameterQuantization(System::Int32 parameterIndex)
{
	return safe_cast<System::Int32>(
		CallHost(Vst2HostCommands::GetParameterQuantization, 0, safe_cast<Vst2IntPtr>(parameterIndex), 0, 0));
}

System::Boolean HostCommandStub::NeedIdle()
{
	return (CallHost(Vst2HostCommands::NeedIdle, 0, 0, 0, 0) != 0);
}

System::IntPtr HostCommandStub::GetPreviousPlugin(System::Int32 pinIndex) // Vst2Plugin*
{
	return System::IntPtr(CallHost(Vst2HostCommands::PluginGetPrevious, 0, safe_cast<Vst2IntPtr>(pinIndex), 0, 0));
}

System::IntPtr HostCommandStub::GetNextPlugin(System::Int32 pinIndex) // Vst2Plugin*
{
	return System::IntPtr(CallHost(Vst2HostCommands::PluginGetNext, 0, safe_cast<Vst2IntPtr>(pinIndex), 0, 0));
}

System::Int32 HostCommandStub::WillReplaceOrAccumulate() // 0=Not Supported, 1=Replace, 2=Accumulate
{
	return safe_cast<System::Int32>(CallHost(Vst2HostCommands::WillReplace, 0, 0, 0, 0));
}

System::Boolean HostCommandStub::SetOutputSampleRate(System::Single sampleRate)
{
	return (CallHost(Vst2HostCommands::SetOutputSampleRate, 0, 0, 0, sampleRate) != 0);
}

Jacobi::Vst::Core::VstSpeakerArrangement^ HostCommandStub::GetOutputSpeakerArrangement()
{
	::Vst2SpeakerArrangement* pArrangement = (::Vst2SpeakerArrangement*)
		CallHost(Vst2HostCommands::GetOutputSpeakerArrangement, 0, 0, 0, 0);

	return Jacobi::Vst::Interop::TypeConverter::ToManagedSpeakerArrangement(pArrangement);
}

System::Boolean HostCommandStub::SetIcon(System::IntPtr icon)
{
	return (CallHost(Vst2HostCommands::SetIcon, 0, 0, icon.ToPointer(), 0) != 0);
}

System::IntPtr HostCommandStub::OpenWindow()    // HWND
{
	return System::IntPtr(CallHost(Vst2HostCommands::WindowOpen, 0, 0, 0, 0));
}

System::Boolean HostCommandStub::CloseWindow(System::IntPtr wnd)
{
	return (CallHost(Vst2HostCommands::WindowClose, 0, 0, wnd.ToPointer(), 0) != 0);
}

System::Boolean HostCommandStub::EditFile(System::String^ xml)
{
	char* pXml = Jacobi::Vst::Interop::TypeConverter::AllocateString(xml);

	try
	{
		return (CallHost(Vst2HostCommands::EditFile, 0, 0, pXml, 0) != 0);
	}
	finally
	{
		Jacobi::Vst::Interop::TypeConverter::DeallocateString(pXml);
	}
}

System::String^ HostCommandStub::GetChunkFile()
{
	Jacobi::Vst::Interop::UnmanagedString pFile(2048);

	CallHost(Vst2HostCommands::GetChunkFile, 0, 0, pFile, 0);

	return Jacobi::Vst::Interop::TypeConverter::CharToString(pFile);
}

Jacobi::Vst::Core::VstSpeakerArrangement^ HostCommandStub::GetInputSpeakerArrangement()
{
	::Vst2SpeakerArrangement* pArrangement = (::Vst2SpeakerArrangement*)
		CallHost(Vst2HostCommands::GetInputSpeakerArrangement, 0, 0, 0, 0);

	return Jacobi::Vst::Interop::TypeConverter::ToManagedSpeakerArrangement(pArrangement);
}

// Throws an InvalidOperationException if the host command stub has not been initialized.
inline void HostCommandStub::ThrowIfNotInitialized()
{
	if (IsInitialized() == false)
	{
		throw gcnew System::InvalidOperationException(
			Jacobi::Vst::Interop::Properties::Resources::HostCommandStub_NotInitialized);
	}
}

}}}} // Jacobi::Vst::Plugin::Interop
